lib/elixir/lib/kernel/typespec.ex

```diff
@@ -129,6 +129,11 @@ defmodule Kernel.Typespec do
       when kind in [:type, :typep, :opaque] do
     {set, bag} = :elixir_module.data_tables(module)
 
+    if invalid_type_ast?(expr) do
+      message = message_for_invalid_type_ast(expr, file)
+      raise(ArgumentError, message: message)
+    end
```

Comment:
What if we merged it all together?

```elixir
defp validate_type_ast!({:::, _, [{type_name, _, [{{:., _, _}, meta, _}]}, _]}) do
  raise ArgumentError, """
  ...
  """
end
```

I don't think we need to fetch the line explicitly - if we raise, it should indicate the line somewhere.